### PR TITLE
Function pointer non-det initialisation in goto-harness [depends-on: #5171]

### PR DESCRIPTION
**Pull request by NlightNFotis**
This is extending the work done in #5171 of which the first commit of this PR belongs to.

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [x] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- [ ] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

File: src/goto-harness/recursive_initialization.cpp [hunk 1]
```diff
@@ -797,3 +801,54 @@ bool recursive_initializationt::needs_freeing(const exprt &expr) const
...
+  }
+
+  code_blockt body{};
+
+  const auto function_pointer_selector =
+    get_fresh_local_symexpr("function_pointer_selector");
+  body.add(
+    code_assignt{function_pointer_selector,
+                 side_effect_expr_nondett{function_pointer_selector.type()}});
+  auto function_pointer_index = 0;
```
Comment (xbauch) [inline src/goto-harness/recursive_initialization.cpp]: `std::size_t` .. this way the deduced type is `int`.

**Review by xbauch — APPROVED** [review]
Looks good.

File: src/goto-harness/recursive_initialization.cpp (hunk 1 above)
Comment (NlightNFotis) [reply src/goto-harness/recursive_initialization.cpp]: Done.

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 825029acf7b833dcff85c26eab549dffa46e6f07).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/135629595

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 60398e78d5a429968028e0357b887a49d1aa1964).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/136133258

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 64a62d4986957f945dc1eaf129c77fde8e205a3f).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/136162388

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: dfadbc87b35ca3d1e151499daba2cd314f40e62b).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/136177150

File: src/goto-harness/function_call_harness_generator.cpp [hunk 2]
```diff
@@ -12,6 +12,7 @@ Author: Diffblue Ltd.
 #include <util/arith_tools.h>
 #include <util/c_types.h>
 #include <util/exception_utils.h>
+#include <util/prefix.h>
```
Comment (xbauch) [inline src/goto-harness/function_call_harness_generator.cpp:15]: do you still need this? (edit: you do, but somewhere in the 4th commit; no need to reorder because of this.)

File: src/goto-harness/recursive_initialization.cpp [hunk 3]
```diff
@@ -264,11 +275,12 @@ irep_idt recursive_initializationt::build_constructor(const exprt &expr)
...
     // the expression name may be needed to decide if expr should be treated as
     // a string
-    expr_name);
+    expr_name,
+    is_nullable);
 
   goto_model.goto_functions.function_map[function_symbol.name].type =
     to_code_type(function_symbol.type);
-  return type_constructor_names[type];
+  return type_constructor_names.at(key);
```
Comment (xbauch) [inline src/goto-harness/recursive_initialization.cpp:323]: Is there a reason why this access is checked but the above are not?

File: src/goto-harness/recursive_initialization.cpp [hunk 4]
```diff
@@ -203,17 +208,23 @@ irep_idt recursive_initializationt::build_constructor(const exprt &expr)
   // void type_constructor_T(int depth_T, T *result_T, int *size);
   optionalt<irep_idt> size_var;
   optionalt<irep_idt> expr_name;
+  bool is_nullable = false;
+  bool has_size_param = false;
   if(expr.id() == ID_symbol)
   {
     expr_name = to_symbol_expr(expr).get_identifier();
+    is_nullable = initialization_config.potential_null_function_pointers.count(
+      expr_name.value());
```
Comment (xbauch) [inline src/goto-harness/recursive_initialization.cpp:257]: Maybe it's idiomatic, but I would prefer `count(..) != 0`

File: src/goto-harness/function_call_harness_generator.cpp [hunk 5]
```diff
@@ -335,6 +347,43 @@ void function_call_harness_generatort::validate_options(
...
+    if(function_pointer_symbol_pointer == nullptr)
+    {
+      throw invalid_command_line_argument_exceptiont{
+        "nullable function pointer `" + id2string(nullable) +
+          "' not found in symbol table",
+        "--" COMMON_HARNESS_GENERATOR_FUNCTION_POINTER_CAN_BE_NULL_OPT};
+    }
+
+    const auto &function_pointer_type =
+      ns.follow(function_pointer_symbol_pointer->type);
```
Comment (xbauch) [inline src/goto-harness/function_call_harness_generator.cpp:394]: Why do we need to follow the type?

File: src/goto-harness/common_harness_generator_options.h [hunk 6]
```diff
@@ -14,13 +14,16 @@ Author: Diffblue Ltd.
...
+#define COMMON_HARNESS_GENERATOR_FUNCTION_POINTER_CAN_BE_NULL_OPT              \
+  "function-pointer-can-be-null"
 
 // clang-format off
 #define COMMON_HARNESS_GENERATOR_OPTIONS                                       \
   "(" COMMON_HARNESS_GENERATOR_MIN_NULL_TREE_DEPTH_OPT "):"                    \
   "(" COMMON_HARNESS_GENERATOR_MAX_NONDET_TREE_DEPTH_OPT "):"                  \
   "(" COMMON_HARNESS_GENERATOR_MIN_ARRAY_SIZE_OPT "):"                         \
   "(" COMMON_HARNESS_GENERATOR_MAX_ARRAY_SIZE_OPT "):"                         \
+  "(" COMMON_HARNESS_GENERATOR_FUNCTION_POINTER_CAN_BE_NULL_OPT "):"           \
```
Comment (xbauch) [inline src/goto-harness/common_harness_generator_options.h:26]: Please add a help description below as well.

File: src/goto-harness/recursive_initialization.cpp [hunk 7]
```diff
@@ -74,6 +74,17 @@ bool recursive_initialization_configt::handle_option(
       COMMON_HARNESS_GENERATOR_MIN_ARRAY_SIZE_OPT, values);
     return true;
   }
+  else if(option == COMMON_HARNESS_GENERATOR_FUNCTION_POINTER_CAN_BE_NULL_OPT)
+  {
+    std::transform(
+      values.begin(),
+      values.end(),
+      std::inserter(
```
Comment (xbauch) [inline src/goto-harness/recursive_initialization.cpp:83]: For the Windows build, you are going to have to `#include <iterator>`

File: src/goto-harness/recursive_initialization.cpp (hunk 4 above)
Comment (hannes-steffenhagen-diffblue) [reply src/goto-harness/recursive_initialization.cpp:257]: I agree that'd be easier to read.

File: src/goto-harness/recursive_initialization.cpp (hunk 3 above)
Comment (hannes-steffenhagen-diffblue) [reply src/goto-harness/recursive_initialization.cpp:323]: Not really, tbh we'd considered refactoring this entire section to avoid multiple lookup of the same key. We only added the `.at` because at some point we were hitting a case where keys were missing from the map.

File: src/goto-harness/function_call_harness_generator.cpp (hunk 5 above)
Comment (hannes-steffenhagen-diffblue) [reply src/goto-harness/function_call_harness_generator.cpp:394]: Symbol types aren't resolved at this point (i.e. in the symbol table). They can't be, otherwise we couldn't correctly print typedefs etc in dump-c.

File: src/goto-harness/function_call_harness_generator.cpp (hunk 2 above)
Comment (hannes-steffenhagen-diffblue) [reply src/goto-harness/function_call_harness_generator.cpp:15]: I think it's worth fixing that.

File: src/goto-harness/recursive_initialization.cpp (hunk 7 above)
Comment (hannes-steffenhagen-diffblue) [reply src/goto-harness/recursive_initialization.cpp:83]: Thanks, I figured it'd be something like that!

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 272477c724339d4cfe65e58d4ba64816f4a6061b).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/136582105

File: src/goto-harness/function_call_harness_generator.cpp [hunk 8]
```diff
@@ -72,10 +72,10 @@ struct function_call_harness_generatort::implt
   void ensure_harness_does_not_already_exist();
   /// Update the goto-model with the new harness function.
   void add_harness_function_to_goto_model(code_blockt function_body);
-  /// declare local variables for each of the parameters of the entry function
+  /// Declare local variables for each of the parameters of the entry function
```
Comment (hannes-steffenhagen-diffblue) [inline src/goto-harness/function_call_harness_generator.cpp:76]: 44509e46ce8d4092a806d3d7bb9c767555ce1932 are good changes, but might be better off in a separate PR (although it’d be annoying to separate out this late I know)

File: src/goto-harness/recursive_initialization.cpp [hunk 9]
```diff
@@ -799,3 +811,57 @@ bool recursive_initializationt::needs_freeing(const exprt &expr) const
...
+  PRECONDITION(can_cast_type<pointer_typet>(result_type.subtype()));
+  const auto &function_pointer_type = to_pointer_type(result_type.subtype());
+  PRECONDITION(can_cast_type<code_typet>(function_pointer_type.subtype()));
+  const auto &function_type = to_code_type(function_pointer_type.subtype());
+
+  std::vector<exprt> targets;
+
+  for(const auto &sym : goto_model.get_symbol_table())
+  {
+    if(sym.second.type == function_type)
```
Comment (hannes-steffenhagen-diffblue) [inline src/goto-harness/recursive_initialization.cpp:920]: We should probably make this work with the function pointer restriction table stuff we have some time.

File: src/goto-harness/recursive_initialization.h [hunk 10]
```diff
@@ -57,7 +58,29 @@ class recursive_initializationt
 {
 public:
   using recursion_sett = std::set<irep_idt>;
-  using type_constructor_namest = std::map<typet, irep_idt>;
+  struct constructor_keyt
+  {
+    typet constructor_type;
+    bool is_nullable;
+    bool has_size_parameter;
+    bool operator<(const constructor_keyt &other) const
```
Comment (hannes-steffenhagen-diffblue) [inline src/goto-harness/recursive_initialization.h:69]: With C++20 we could just default these ;)

File: src/goto-harness/recursive_initialization.h [hunk 11]
```diff
@@ -165,19 +188,29 @@ class recursive_initializationt
...
   /// \return name of the constructor function
   irep_idt build_constructor(const exprt &expr);
 
+  /// Constructor for function pointers.
+  /// \param result: symbol of the result parameter
+  /// \param is_nullable: if the function pointer can be null
+  /// \return the body of the constructor
+  code_blockt build_function_pointer_constructor(
+    const symbol_exprt &result,
+    bool is_nullable);
```
Comment (hannes-steffenhagen-diffblue) [inline src/goto-harness/recursive_initialization.h:217]: I’m not extremely excited about having to pass this `is_nullable` argument down all the time, but I can’t think of anything super great either.

File: src/goto-harness/function_call_harness_generator.cpp [hunk 12]
```diff
@@ -81,6 +81,16 @@ struct function_call_harness_generatort::implt
   void call_function(
     const code_function_callt::argumentst &arguments,
     code_blockt &function_body);
+  /// For function parameters that are pointers to functions we want to
+  /// be able to specify whether or not they can be NULL. To disambiguate
+  /// this specification from that for a global variable of the same name,
+  /// we prepend the name of the function to the parameter name. However,
+  /// what is actually being initialised in the implementation is not the
+  /// parameter itself, but a correspond function argument (local variable
```
Comment (hannes-steffenhagen-diffblue) [inline src/goto-harness/function_call_harness_generator.cpp]: correspond(ing)

**Review by hannes-steffenhagen-diffblue — APPROVED** [review]
